packages/button/src/react/index.tsx

```diff
@@ -136,109 +132,111 @@ interface BaseButtonProps {
-    const anchorProps = rest as HTMLAttributes<HTMLAnchorElement>
-
-    return (
-      <a
-        ref={ref as React.Ref<HTMLAnchorElement>}
-        {...glamorStyle}
-        {...anchorProps}
-        onClick={disabled ? undefined : anchorProps.onClick}
-        style={style}
-      >
-        {iconEl}
-        {labelEl}
-      </a>
-    )
-  } else {
-    const buttonProps = rest as HTMLAttributes<HTMLButtonElement>
-    delete (buttonProps as any).download
-
-    return (
-      <button
-        disabled={disabled || loading}
-        ref={ref as React.Ref<HTMLButtonElement>}
-        {...glamorStyle}
-        {...buttonProps}
-        style={style}
-      >
-        {iconEl}
-        {labelEl}
-      </button>
+const Button = React.forwardRef<ButtonElement, ButtonProps>(
+  (props, forwardedRef) => {
+    const {
+      appearance = vars.appearances.primary,
+      children,
+      disabled = false,
+      icon,
+      iconAlign = vars.iconAligns.left,
+      loading = false,
+      size = vars.sizes.medium,
+      ...rest
+    } = props
+
+    const iconOnly = Children.count(children) <= 0
+
+    const themeName = useTheme()
+
+    const ref = React.useRef<HTMLAnchorElement | HTMLButtonElement>()
+    React.useImperativeHandle(forwardedRef, () => ref.current)
+
+    const nonLoadingWidth = React.useMemo(() => {
+      if (loading && ref && ref.current) return ref.current.offsetWidth
+    }, [loading, ref])
+    const isLoadingWithNoText = !!nonLoadingWidth
+
+    const glamorStyle = styles.button({
+      appearance,
+      disabled,
+      icon,
+      iconAlign,
+      iconOnly,
+      size,
+      themeName
+    })
+    const style = {
+      ...(props.style || {}),
+      ...(isLoadingWithNoText && { width: nonLoadingWidth })
+    }
+    const iconEl = renderIcon({
+      appearance,
+      icon,
+      iconAlign,
+      iconOnly,
+      isLoadingWithNoText,
+      loading,
+      size,
+      themeName
+    })
+    const labelEl = !isLoadingWithNoText && (
+      <span {...styles.text()}>{children}</span>
     )
+
+    if ('href' in props && typeof props.href === 'string') {
+      const anchorProps = rest as HTMLPropsFor<'a'>
+
+      return (
+        <a
+          ref={ref as React.Ref<HTMLAnchorElement>}
+          {...glamorStyle}
+          {...anchorProps}
+          onClick={disabled ? undefined : anchorProps.onClick}
+          style={style}
+        >
+          {iconEl}
+          {labelEl}
+        </a>
+      )
+    } else {
+      const buttonProps = rest as HTMLPropsFor<'button'>
+      delete (buttonProps as any).download
+
+      return (
+        <button
+          disabled={disabled || loading}
+          ref={ref as React.Ref<HTMLButtonElement>}
+          {...glamorStyle}
+          {...buttonProps}
+          style={style}
+        >
+          {iconEl}
+          {labelEl}
+        </button>
+      )
+    }
   }
-}) as ButtonComponent
+) as ButtonComponent & ButtonStatics
```

Comment:
I also like that this approach uses the React typings directly w/o needing our wrapper.